lib/js/jslib.js

```diff
@@ -1,187 +1,255 @@
+        var cur = list._head;
+        f(cur);
+        list = list._tail;
+      }
+    },
 
-function _lsTake (n, xs) {
-  var arr = [ ];
-  while (xs !== null && n > 0) {
-    arr.push(xs._head);
-    xs = xs._tail;
-    --n;
-  }
-  return _lsFromArray(arr);
-}
-const lsTake = _lsTake;
+    map: function(list, f) {
+      var cur = LINKEDLIST.singleton(null);
+      var res = cur;
+      while(!LINKEDLIST.isNil(list)) {
+        cur._tail = LINKEDLIST.Cons(f(list._head),Nil);
+        cur = cur.tail;
+        list = list._tail;
+      }
+      return res._tail;
+    },
 
-function _lsDrop (n, xs) {
-  while (xs !== null && n > 0) {
-    xs = xs._tail;
-    --n;
-  }
-  return xs;
-}
-const lsDrop = _lsDrop;
+    mapFromArray: function(array, f) {
+      var cur = LINKEDLIST.singleton(null);
+      var res = cur;
+      for(var i = 0; i < array.length; i++) {
+        cur._tail = LINKEDLIST.Cons(f(array[i]),Nil);
+        cur = cur.tail;
+        list = list._tail;
+      }
+      return res._tail;
+    },
 
-function _lsLength (xs) {
-  var out = 0;
-  while (xs !== null) {
-    out += 1;
-    xs = xs._tail;
-  }
-  return out;
-}
-const lsLength = _lsLength;
+    toArray: function(list) {
+      var res = [];
+      var cur = list;
+      while(!LINKEDLIST.isNil(cur)) {
+        res.push(cur._head);
+        cur = cur._tail;
+      }
+      return res;
+    },
 
-function _lsHead (v) { return v === null ? _error('head') : v._head; } // inline?
-const lsHead = _lsHead;
 
-function _lsTail (v) { return v === null ? _error('tail') : v._tail; }
-const lsTail = _lsTail;
+    snoc: function(xs, x) { return LINKEDLIST.append(xs, LINKEDLIST.singleton(x)); },
 
-function _lsLast (xs) {
-  if (xs === null) {
-    return _error('last');
-  }
-  var out = xs._head;
-  while (xs !== null) {
-    out = xs._head;
-    xs = xs._tail;
-  }
-  return out;
-}
-const lsLast = _lsLast;
+    singleton: function(head) { return LINKEDLIST.Cons(head,Nil); },
 
-function _lsRevAppend (xs, ys) {
-  var out = ys;
-  while (xs !== lsNil) {
-    out = _lsCons(_lsHead(xs), out);
-    xs = _lsTail(xs);
-  }
-  return out;
-}
-const lsRevAppend = _lsRevAppend;
+    nilF: function() {return Nil},
 
-function _lsAppend (xs, ys) {
-  if (xs === null) {
-    return ys;
-  }
-  const rootEl = _lsCons(xs._head, lsNil);
-  let curr = rootEl;
-  xs = xs._tail;
-  while (xs !== null) {
-    curr._tail = _lsCons(xs._head, lsNil);
-    xs = xs._tail;
-    curr = curr._tail;
-  }
-  curr._tail = ys;
-  return rootEl;
-}
-const lsAppend = _lsAppend;
+    lsFromArray: function(arr) {
+      var out = Nil;
+      for (let i = arr.length - 1; i >= 0; --i) {
+        out = LINKEDLIST.Cons(arr[i], out);
+      }
+      return out;
+    },
 
-function _lsAt (xs, i) {
-  var out;
-  while (xs !== null && i >= 0) {
-    out = xs._head;
-    xs = xs._tail;
-    --i;
-  }
-  // FIXME should check i / index out of bounds here
-  return out;
-}
-const lsAt = _lsAt;
+    take: function (n, xs) {
```

Comment:
2) With the new (uniform) list representation I think this would be defined in the prelude rather as a built-in function. To the best of my knowledge the Links compiler does not depend on `take`.